### PR TITLE
BUG: Fix floyd_warshall to support zero-weight edges

### LINKED ISSUE #10609: Graph shortest path with Floyd-Warshall removes explicit zeros.
My issue concerns `scipy.sparse.csgraph.shortest_path`, more specifically the Floyd-Warshall solver, which does not deal with **explicit zeros** as the other algorithms (they are removed [here](https://github.com/scipy/scipy/blob/master/scipy/sparse/csgraph/_shortest_path.pyx#L343)). 

The following (slightly convoluted) example highlights the different handling of explicit zeros.

#### Reproducing code example:
```py
import numpy as np
import scipy.sparse.csgraph
import sklearn.neighbors

examples = [
    # no infinity
    # (the undirected graph has one connected component)
    np.array([0, 1, 2, 3, 3])[:, None],
    # infinity only for 'FW' solver
    # (the undirected graph has one connected component)
    np.array([0, 1, 2, 3, 3, 3])[:, None],
    # infinity for all solvers
    # (the undirected graph has two connected components)
    np.array([0, 1, 2, 4, 4, 4])[:, None],
]

for X in examples:
    graph = sklearn.neighbors.kneighbors_graph(X, n_neighbors=2,
                                               mode='distance')
    print(scipy.sparse.csgraph.connected_components(graph)[0], end='\t')

    for method in ['FW', 'D', 'BF', 'J']:
        paths = scipy.sparse.csgraph.shortest_path(graph, method=method,
                                                   directed=False)
        print(np.any(np.isinf(paths)), end='\t')

    print('')
```

#### Print message:

```
1	False	False	False	False	
1	True	False	False	False	
2	True	True	True	True
```
#### Explanations:

Explicit zeros arise when you have duplicate points in your graph, thus having a connection with a distance of zero, which is different from the implicit zero (meaning no graph connection). When building a k-nearest neighbors graph, if some points are duplicated **more** than the number of neighbors k, then these points have **only** zeros in the set of explicit values, and using explicit zeros as if they were implicit zeros effectively remove the points from the connected graph. In particular, these points may still be connected to the rest if they are chosen as neighbors of another point (with `directed=False`).

#### Scipy/Numpy/Python version information:

1.2.1 1.16.4 sys.version_info(major=3, minor=7, micro=3, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by huizew**
#### Reference issue

Closes #10609

#### What does this implement/fix?

1. Fix `floyd_warshall` by moving the "set non-edges to infinity" step out of the `_floyd_warshall` helper function, as suggested by @perimosocordiae

2. Created test cases that have zero-weight edges, for both directed and undirected graphs.

### CODE REVIEW

**Comment by huizew** [comment]
Don't understand why CircleCI build_docs tests failed ([Details](https://circleci.com/gh/scipy/scipy/14909?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)). Any suggestions?

File: scipy/sparse/csgraph/tests/test_shortest_path.py
Comment (tylerjereddy) [inline scipy/sparse/csgraph/tests/test_shortest_path.py]: could condense to  a single test parametrized over `directed_sparse_zero_SP, directed_sparse_zero_SP` and `methods`. Can't fault you for using the same test structure as elsewhere in the module though

**Review by tylerjereddy — COMMENTED** [review]
For the CI doc build failure, we've already fixed that in master a few weeks ago so you likely just need to rebase your feature branch on the latest master & force push. I see that you are using the master branch on your fork instead of a separate feature branch name, which can also cause confusion--usually best to use a name unique from master so you can sync your master with upstream and develop feature branches independent of that.

I made a small testing suggestion, but apart from that looks promising with regression tests already here. Will wait for CJ to chime in.

**Comment by huizew** [comment]
@tylerjereddy Thanks for the review! I just realized that `test_undirected` was different from `test_directed` because it was intended to test the `directed` option, as described [here](https://docs.scipy.org/doc/scipy/reference/sparse.csgraph.html#directed-vs-undirected). So I modified my `test_undirected_sparse_zero` to test this option as well.

As for the CI doc build failure, I just forked scipy master yesterday, and I just check that they are up to date. Am I missing something? Or is there anything I can do to fix this?

Thanks!

**Comment by tylerjereddy** [comment]
Your fork is [668 commits behind master](https://github.com/huizew/scipy/commits/master), and doesn't include anything from the last few months

**Comment by huizew** [comment]
My bad. Fixed now. Thanks! @tylerjereddy

**Review by perimosocordiae — APPROVED** [review]

**Comment by perimosocordiae** [comment]
It took me a while to verify that the interaction between `validate_graph` and sparse matrices here, but this looks good to me. I'll let @tylerjereddy take another look before merging.

**Review by tylerjereddy — APPROVED** [review]
I confirmed that the MWE provided in the cognate issue still reproduces the problem on latest master branch, then locally rebased this feature branch on latest master & confirmed the behavior was adjusted appropriately so that the MWE passes.

Combined with CJ's +1, and the fact that no old tests were modified & only new tests added -- merging in, thanks.

**Pull request merged**
